internal/app/orchestrator/orchestrator.go

```diff
@@ -95,16 +114,183 @@ func New(ctx context.Context,
 }
 
-func (c *orchestrator) onCacheEvicted(key string, resource cache.Resource) {
-	// TODO implement.
+// watchResponse is intended to be called in a goroutine, to receive incoming
+// responses and fan out to downstream clients.
+func (o *orchestrator) watchUpstream(
+	ctx context.Context,
+	aggregatedKey string,
+	responseChannel <-chan *discovery.DiscoveryResponse,
+	done <-chan bool,
+	shutdown func(),
+) {
+	for {
+		select {
+		case x, more := <-responseChannel:
+			if !more {
+				// A problem occurred fetching the response upstream, log and
+				// return the most recent cached response, so that the
+				// downstream will reissue the discovery request.
+				o.logger.With("key", aggregatedKey).Error(ctx, "upstream error")
+			} else {
+				// Cache the response.
+				_, err := o.cache.SetResponse(aggregatedKey, *x)
+				if err != nil {
+					// If we fail to cache the new response, log and return the old one.
+					o.logger.With("err", err).With("key", aggregatedKey).
+						Error(ctx, "Failed to cache the response")
+				}
+			}
+
+			// Get downstream watches and fan out.
+			// We retrieve from cache rather than directly fanning out the
+			// newly received response because the cache does additional
+			// resource serialization.
+			cached, err := o.cache.Fetch(aggregatedKey)
+			if err != nil {
+				o.logger.With("err", err).With("key", aggregatedKey).Error(ctx, "cache fetch failed")
+				// Can't do anything because we don't know who the watches
+				// are. Drop the response.
+			} else {
+				if cached == nil || cached.Resp == nil {
+					// If cache is empty, there is nothing to fan out.
+					if !more {
+						// Warn. Benefit of the doubt that this is the first request.
+						o.logger.With("key", aggregatedKey).
+							Warn(ctx, "attempted to fan out with no cached response")
+					} else {
+						// Error. Sanity check. Shouldn't ever reach this.
+						o.logger.With("key", aggregatedKey).
+							Error(ctx, "attempted to fan out with no cached response")
+					}
+				} else {
+					// Goldenpath.
+					o.fanout(cached.Resp, cached.Requests)
+				}
+			}
+		case <-done:
+			// Exit when signaled that the stream has closed.
+			shutdown()
+			return
+		}
+	}
+}
+
+// fanout pushes the response to the response channels of all open downstream
+// watches in parallel.
+func (o *orchestrator) fanout(resp *cache.Response, watchers map[*gcp.Request]bool) {
+	for watch := range watchers {
+		go func(watch *gcp.Request) {
+			if channel, ok := o.downstreamResponseMap.get(watch); ok {
+				channel <- convertToGcpResponse(resp, *watch)
+			}
+		}(watch)
+	}
+}
+
+// onCacheEvicted is called when the cache evicts a response due to TTL or
+// other reasons. When this happens, we need to clean up open streams.
+// We shut down both the downstream watches and the upstream stream.
+func (o *orchestrator) onCacheEvicted(key string, resource cache.Resource) {
+	o.downstreamResponseMap.deleteAll(resource.Requests)
```

Comment:
i think we discussed that removing all sidecars will cause a thundering herd problem. We need a way to open a new stream on upstream only when ttl expires.

Reply:
Chatted offline. We're going to add a TODO to issue a retry on the upstream rather than closing the downstream connections.

Reply:
opened https://github.com/envoyproxy/xds-relay/issues/71